### PR TITLE
Implement cluster CLI commands

### PR DESCRIPTION
**Pull request by jadeallenx**
Implement the following commands using riak_cli:
- cluster status (improved member-status)
- partition-count
- partitions
- partition id=<id>
- partition index=<index>
- members

TODO: After merge, some code clean up and boundary checking.

Depends on basho/riak_ee#295

### CODE REVIEW

File: src/riak_core_cluster_cli.erl [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+      {Node, Status} <- AllStatus ],
+
+    Table = riak_cli_status:table(Rows),
+
+    T0 = riak_cli_status:text("---- Cluster Status ----"),
+    T1 = riak_cli_status:text(io_lib:format("Ring ready: ~p~n", [element(2, RingStatus)])),
+    T2 = riak_cli_status:text(
+           "Key: (C) = Claimant; availability marked with '!' is unexpected"),
+    Out = riak_cli_status:alert([T0,T1,Table,T2]),
+    riak_cli:print([Out]).
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: I've been thinking about this a bit. I didn't really intend for implementers to constantly call `riak_cli:print/1`. I intended for it to be called automatically and use the proper writer module given global --csv flags etc..., but of course I didn't implement it that way. I spent a lot of time focusing on config the last few weeks and if you look in riak_cli.erl you'll see that the config ops (show, describe, set) wrap the output in a print statement, and the callbacks themselves return errors tuples or status types. Returning errors works fine for config because it's all internal and riak_cli_error knows how to parse the riak_cli specific errors and generate alerts from them. However, since that requires modifying riak_cli itself, outside users of riak_cli should only ever return status types in callbacks. 

This was a mistake in implementation that I will fix on Monday. We can then modify each PR to do the right thing and only return a status(). In your case it will just be removing the riak_cli:print/1 calls. Hopefully the same elsewhere.

File: src/riak_core_cluster_cli.erl [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+id_out1(id, Id, Ring, RingSize) when Id < RingSize ->
+    Idx = partition_id_to_hash(Id, RingSize),
+    Owner = riak_core_ring:index_owner(Ring, Idx),
+    T = riak_cli_status:text(
+          io_lib:format("Partition id: ~p -> index: ~p~n\t(owner: ~p)~n", 
+              [Id, partition_id_to_hash(Id, RingSize), Owner])),
+    riak_cli:print([T]);
+id_out1(id, Id, _Ring, _RingSize) ->
+    {error, ["Id ", Id, " is invalid."]}.
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: This doesn't mean anything to riak_cli. I'm not sure what exactly you are trying to do here. When we fix things up it should return an alert status.

File: src/riak_core_cluster_cli.erl [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+
+member_output(L) ->
+    O = [ riak_cli_status:text(atom_to_list(N)) || {N, _S} <- L ],
+    riak_cli:print(O).
+
+get_status() ->
+    {ok, Ring} = riak_core_ring_manager:get_my_ring(),
+    lists:keysort(2, riak_core_ring:all_member_status(Ring)).
+
+%%% FIXME! -> REMOVE AFTER MERGE
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl:287]: Is this John's code that needs to be merged into the new integration branch?

File: src/riak_core_cluster_cli.erl [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+                   partitions_usage(), partition_usage(),
+                   members_usage()
+                  ]).
+
+%%%
+%% Cluster status
+%%%
+
+status_register() ->
+    [
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: Just my personal style opinion, but I don't like the brackets on separate lines in erlang. I prefer something like this, but It's your decision.

``` erlang
       [["riak-admin", "cluster", "status"], % Cmd
        [],                                  % KeySpecs
        [],                                  % FlagSpecs
        fun status/2].                       % Implementation callback
```

File: src/riak_core_cluster_cli.erl (hunk 1 above)
Comment (jadeallenx) [inline src/riak_core_cluster_cli.erl]: OK, gotcha. Now that basho/riak_cli#14 has been merged, I will fix it up.

File: src/riak_core_cluster_cli.erl (hunk 2 above)
Comment (jadeallenx) [inline src/riak_core_cluster_cli.erl]: OK, will fix.

File: src/riak_core_cluster_cli.erl (hunk 3 above)
Comment (jadeallenx) [inline src/riak_core_cluster_cli.erl:287]: Yes. It is, but I couldn't find a clean commit to cherry pick so I just cut'n'pasted; my plan was to rework these calls once we have a branch with all the things.

File: src/riak_core_cluster_cli.erl (hunk 4 above)
Comment (jadeallenx) [inline src/riak_core_cluster_cli.erl]: OK. I don't have a strong preference, but I found it avoided merged conflicts sometimes. I will fix it up.

**Comment by jadeallenx** [comment]
@andrewjstone I've pushed a new commit responding to your code review.

File: src/riak_core_cluster_cli.erl [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+format_status(Node, Status, Ring, RingStatus) ->
+    {Claimant, _RingReady, Down, MarkedDown, Changes} = RingStatus,
+    [{node, is_claimant(Node, Claimant)},
+     {status, Status},
+     {avail, node_availability(Node, Down, MarkedDown)},
+     {ring, claim_percent(Ring, Node)},
+     {pending, future_claim_percentage(Changes, Ring, Node)}
+    ].
+
+is_claimant(Node, Node) ->
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl:114]: This is totally a bikeshed comment and won't hold up a +1 of this PR. I was just thinking that we should align this better. Either put the (c) after the Node name, since it's left aligned, or  put 3 spaces before the non-claimant nodes so it is right - aligned.

File: src/riak_core_cluster_cli.erl [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    Rows = [ format_status(Node, Status, Ring, RingStatus) ||
+      {Node, Status} <- AllStatus ],
+
+    Table = riak_cli_status:table(Rows),
+
+    T0 = riak_cli_status:text("---- Cluster Status ----"),
+    T1 = riak_cli_status:text(io_lib:format("Ring ready: ~p~n", [element(2, RingStatus)])),
+    T2 = riak_cli_status:text(
+           "Key: (C) = Claimant; availability marked with '!' is unexpected"),
+    [riak_cli_status:alert([T0,T1,Table,T2])].
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: Why did you wrap these elements in alerts?

File: src/riak_core_cluster_cli.erl [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+partitions_output(Node) ->
+    {ok, Ring} = riak_core_ring_manager:get_my_ring(),
+    RingSize = riak_core_ring:num_partitions(Ring),
+    {Primary, Secondary, Stopped} = riak_core_status:partitions(Node, Ring),
+    T0 = riak_cli_status:text(io_lib:format("Partitions owned by ~p:", [Node])),
+    Rows = generate_rows(RingSize, primary, Primary) 
+           ++ generate_rows(RingSize, secondary, Secondary)
+           ++ generate_rows(RingSize, stopped, Stopped),
+    Table = riak_cli_status:table(Rows),
+    [riak_cli_status:alert([T0, Table])].
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: Why use an alert here?

File: src/riak_core_cluster_cli.erl (hunk 6 above)
Comment (jadeallenx) [inline src/riak_core_cluster_cli.erl]: I think it's just my general misunderstanding of how status types ought to be used, but easily fixed.

File: src/riak_core_cluster_cli.erl (hunk 7 above)
Comment (jadeallenx) [inline src/riak_core_cluster_cli.erl]: Will fix.

**Comment by jadeallenx** [comment]
Ping @andrewjstone

File: src/riak_core_cluster_cli.erl [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+
+is_claimant(Node, Node) ->
+    " (C) " ++ atom_to_list(Node) ++ " ";
+is_claimant(Node, _Other) ->
+    "     " ++ atom_to_list(Node) ++ " ".
+
+node_availability(Node, Down, MarkedDown) ->
+    case {lists:member(Node, Down), lists:member(Node, MarkedDown)} of
+        {false, false} -> "  up   "; % put common case at the top
+          {true, true} -> " down  ";
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: Can you line this clause up please?

File: src/riak_core_cluster_cli.erl [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+%%%
+%% cluster partition-count
+%%%
+
+partition_count_register() ->
+    [
+     ["riak-admin", "cluster", "partition-count"], % Cmd
+     [],                                           % KeySpecs
+     [{node, [{shortname, "n"}, {longname, "node"}, 
+              {typecast, fun list_to_atom/1}]}],   % FlagSpecs
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: This should change to use `riak_cli_typecast:to_node/1`. Otherwise you end up with the following: 

```
Andrews-MacBook-Pro:riak_ee ajs$ dev/dev1/bin/riak-admin cluster partition-count --node=dev2
+----+----------+-----+
|node|partitions| pct |
+----+----------+-----+
|dev2|    0     |  0.0|
+----+----------+-----+
```

File: src/riak_core_cluster_cli.erl [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+
+%%%
+%% cluster partitions
+%%%
+
+partitions_register() ->
+    [["riak-admin", "cluster", "partitions"],      % Cmd
+     [],                                           % KeySpecs
+     [{node, [{shortname, "n"}, {longname, "node"},
+              {typecast, fun list_to_atom/1}]}],   % FlagSpecs
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: Use: `riak_cli_typecast:to_node/1`. Yep, I realize this is all my fault.

File: src/riak_core_cluster_cli.erl [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+            "riak-admin cluster partition index=22835963083295358096932575511191922182123945984\n\n",
+            "Returns the id or index for the specified index or id.\n"],
+    [["riak-admin", "cluster", "partition"], Text].
+
+partition([{index, Index}], []) when Index >= 0 ->
+    id_out(index, Index);
+partition([{id, Id}], []) when Id >= 0 ->
+    id_out(id, Id);
+partition([{Op, Value}], []) ->
+    [riak_cli_status:text(["ERROR: The given value ", integer_to_list(Value),
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: So, I clearly need to write better docs as I've learned the last couple days. Errors should actually be wrapped in alerts. That way, eventually, we can bold or make them red, etc... We also can build separate error csv files.

File: src/riak_core_cluster_cli.erl [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+          io_lib:format("Partition index: ~p -> id: ~p~n\t(owner: ~p)", 
+              [Index, hash_to_partition_id(Index, RingSize), Owner]));
+id_out1(id, Id, Ring, RingSize) when Id < RingSize ->
+    Idx = partition_id_to_hash(Id, RingSize),
+    Owner = riak_core_ring:index_owner(Ring, Idx),
+    riak_cli_status:text(
+          io_lib:format("Partition id: ~p -> index: ~p~n\t(owner: ~p)~n", 
+              [Id, partition_id_to_hash(Id, RingSize), Owner]));
+id_out1(id, Id, _Ring, _RingSize) ->
+    riak_cli_status:text(["ERROR: Id ", integer_to_list(Id), " is invalid."]).
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: Same as above, wrap this in an alert please.

File: src/riak_core_cluster_cli.erl [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+    {ok, Ring} = riak_core_ring_manager:get_my_ring(),
+    RingSize = riak_core_ring:num_partitions(Ring),
+    [id_out1(InputType, Number, Ring, RingSize)].
+
+%% XXX FIXME: Validation of id - test for boundary condition and that it's lower
+%% than the ringsize - 1 partition id
+id_out1(index, Index, Ring, RingSize) ->
+    Owner = riak_core_ring:index_owner(Ring, Index),
+    riak_cli_status:text(
+          io_lib:format("Partition index: ~p -> id: ~p~n\t(owner: ~p)",
```
Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: I don't really like the `\t` here. I prefer things left-aligned. Even if you put the id on the next line with the owner it would work left-aligned. I will open this bikeshed up to @macintux @aberghage.

Currently output looks like this:

```
Andrews-MacBook-Pro:riak_ee ajs$ dev/dev1/bin/riak-admin cluster partition index=867766597165223607683437869425293042920709947392
Partition index: 867766597165223607683437869425293042920709947392 -> id: 38
    (owner: 'dev2@127.0.0.1')
```

Comment (andrewjstone) [inline src/riak_core_cluster_cli.erl]: We should do the same thing with partition id=XX also.

**Comment by jadeallenx** [comment]
@andrewjstone think 3e28706 is a winner :)

**Comment by andrewjstone** [comment]
I'm seeing the following behavior, but I believe it should be fixed in riak_cli. In this case `["riak-admin", "cluster", "partition"]` has both a command and usage point. If riak_cli knew that the command required more than 1 argument it could automatically show usage if no arguments were given. This would require each user to change their specs to inform riak_cli of the number of required arguments. Alternatively, we could always show the usage if a client returns `{error, usage}` or `{error, usage, Status}`. That's probably a more robust solution for cases like these.

```
Andrews-MacBook-Pro:riak_ee ajs$ dev/dev1/bin/riak-admin cluster partition
RPC to 'dev1@127.0.0.1' failed: {'EXIT',
                                 {function_clause,
                                  [{riak_core_cluster_cli,partition,
                                    [[],[]],
                                    [{file,"src/riak_core_cluster_cli.erl"},
                                     {line,218}]},
                                   {riak_cli,run,1,
                                    [{file,"src/riak_cli.erl"},{line,70}]},
                                   {rpc,'-handle_call_call/6-fun-0-',5,
                                    [{file,"rpc.erl"},{line,205}]}]}}
Andrews-MacBook-Pro:riak_ee ajs$ dev/dev1/bin/riak-admin cluster partition -h
Usage: riak-admin cluster partition id=0
riak-admin cluster partition index=22835963083295358096932575511191922182123945984

Returns the id or index for the specified index or id.
```

**Comment by andrewjstone** [comment]
:+1:  3e28706

Great work @mrallen1

**Comment by andrewjstone** [comment]
@borshop merge
